src/NBitcoin/Block.cs

```diff
@@ -106,6 +106,11 @@ public virtual void ReadWrite(BitcoinStream stream)
         #endregion
 
         public virtual uint256 GetHash()
+        {
+            return this.GetHash(null);
```

Comment:
If GetHash's input param is nullable why not just have 

```
public virtual uint256 GetHash(ConsensusFactory consensusFactory = null)
```

and from `GetHash()` call that overload?

Reply:
Because there are some places in the code (in Moq methods) that are using GetHash() in an expression and expressions don't play well with optional arguments.
So I was receiving an error saying `An expression tree may not contain a call or invocation that uses optional arguments`.

To illustrate what this is, let me show you a small example:
```
void Main()
{
	MyClass myclass1 = new MyClass();	
	Console.WriteLine(myclass1.MethodWithExpression(c => myclass1.Method1()));		
}

class MyClass
{	
	public string Method1(string first = null)
	{
		return first;
	}

	public object MethodWithExpression(Expression<Func<string, string>> myExpression)
	{		
		return myExpression.Body;
	}
}

```
This piece of code will show the error I described above around `myclass1.Method1()`.
I think the reason why it's not supported is because since the Expression is just representing the lambda expression without compiling it, it doesn't want to deal with creating a tree that would accomodate all overloads (as the compiler would do for a normal method with optional arguments).